mono/metadata/threads.c

```diff
@@ -2085,16 +2086,59 @@ ves_icall_System_Threading_WaitHandle_Wait_internal (gpointer *handles, gint32 n
 	// Allocate handle outside of the loop.
 	MonoExceptionHandle exc = MONO_HANDLE_NEW (MonoException, NULL);
 
+	// Duplicate objects are not allowed for WaitAll, but are ok with WaitAny.
+	// There are the following implementation choices:
+	//  - qsort and look for adjacent equal, or bubble-sort-like and look for equal
+	//  - handle-based or object-based
+	//  - check for duplicate earlier or later
+	//
+	// Multiple combinations exist.
+	//
+	// CoreCLR/Windows: handle-based, sorted, checking after an error
+	// CoreCLR/Unix: object-based, bubble-sort-like
+	// Mono/Unix: object-based, sorted
+	// Mono/Windows: handle-based, sorted
+	//
+	// MSDN documentation is unclear as to handle-based or object-based -- probably object-based.
+	// For .NET to be object-based in a thin Win32-based implementation is impossible.
+	//
+	// Note that the lowest array index signaled is to be returned, so
+	// operating on the sorted data is incorrect, or needs an indirection to "unsort".
+
 	for (;;) {
 #ifdef HOST_WIN32
+		DWORD win32WaitResult;
 		MONO_ENTER_GC_SAFE;
-		ret =	(numhandles != 1)
-			? mono_w32handle_convert_wait_ret (mono_win32_wait_for_multiple_objects_ex(numhandles, handles, waitall, timeoutLeft, TRUE), numhandles)
-			: mono_w32handle_convert_wait_ret (mono_win32_wait_for_single_object_ex (handles [0], timeoutLeft, TRUE), 1);
+		win32WaitResult = (numhandles != 1)
+			? mono_win32_wait_for_multiple_objects_ex (numhandles, handles, waitall, timeoutLeft, TRUE)
+			: mono_win32_wait_for_single_object_ex (handles [0], timeoutLeft, TRUE);
 		MONO_EXIT_GC_SAFE;
+		// Duplicate detection here is based on CoreCLR.
+		// In particular, it is triggered by an error, and it is handled-based
+		// instead of object-based.
+		// See coreclr/src/vm/threads.cpp CheckForDuplicateHandles and its use.
+		if (win32WaitResult == WAIT_FAILED
```

Comment:
I'm not 100% sure of whether we need to keep this weird split on error detection.
Can't we just have mono_win32_wait_for_multiple_objects_ex return a `MONO_W32HANDLE_WAIT_RET_FAILED_DUPLICATES`?

@luhenry what do you think?

Reply:
The existing win32 layer is very thin & so I left it alone after brief consideration, but yes that might be a better structure.

Reply:
Tests in mono/tests takes a LOT longer to run compared to the unit tests in corlib.

Plus if you're making changes that are easily visible to C#, you should be running that test suite anyways.